### DISCUSSION TITLE
Plans for Epiphany 44+?

### DISCUSSION THREAD

**Discussion by bobby285271** [elementary/browser#82] [General]
Sorry if I am providing false positive here since NixOS does not start packaging GNOME 44 yet and all I say here are some guessing :sweat_drops:

I think Epiphany 44 is ported to GTK4 and libadwaita, I installed [org.gnome.Epiphany.Devel.flatpakref](https://nightly.gnome.org/repo/appstream/org.gnome.Epiphany.Devel.flatpakref) and it looks like it is using the Adwaita stylesheet instead of the elementary stylesheet. ***I assume*** when we package Epiphany 44 in Nixpkgs it should look the same way.

![](https://user-images.githubusercontent.com/20080233/218913653-e8d360bc-0af1-47f8-80a5-ccb2e6abb316.png)

I personally prefer the app to use the elementary stylesheet on Pantheon desktop when possible. Is there some plan for elementary to do that? For example:

1. Do something conditionally using upstream's [`is_desktop_pantheon()`](https://gitlab.gnome.org/search?search=is_desktop_pantheon&nav_source=navbar&project_id=1906&group_id=8&search_code=true&repository_ref=master)
1. Use `Gtk.Application` instead of `Adw.Application`
1. ...

Or is there some suggestions on how we should take care of this in NixOS (we are shipping Epiphany on default Pantheon installations. we probably can't pin Epiphany to an old version specifically for Pantheon, but we can ship some Pantheon specific patches when needed as long as it is shipped by elementary and it applies cleanly on latest Epiphany - [we are already doing that](https://github.com/NixOS/nixpkgs/blob/179bff391b21001028b6cc3a3ff6f6de44b897a8/pkgs/desktops/gnome/core/epiphany/default.nix#L48))?

**Comment by mcatanzaro** [comment]
CC @alatiera

**Comment by danirabbit** [comment]
I imagine for working around `Adw.Application`'s stylesheet hardcoding that we'll need to carry a patch set, unless we can ask upstream for a conditional compile maybe? It's not something I've dived into just yet :sweat_smile:

**Reply by mcatanzaro** [reply]
I think the intent is that Epiphany supports both Adwaita and Granite themes, even though it only uses libadwaita and not Granite. Some things might need patching but you can do that upstream. Epiphany's upstream flatpak still bundles the elementary stylesheets, so it's understood there will be a some elementary-specific code.

**Reply by danirabbit** [reply]
@Exalm might have more insight into this, but I think when you use Adw.Application it forces the Adwaita stylesheet. Is there a way in Adwaita to change that?

**Reply by Marukesu** [reply]
I don't think we can change the stylesheet libadwaita loads or stop it from loading it, however, we can override it by using a StyleProvider with the elementary stylesheet and a higher priority than the libadwaita one, listening the settings portal for accent changes.

**Reply by alice-mkh** [reply]
There technically is, but it's unsupported and I can't guarantee it will stay (`GTK_THEME` env var).

> we can override it by using a StyleProvider with the elementary stylesheet and a higher priority than the libadwaita one

That would be an absolute mess, you'll get both styles conflicting.

Generally ephy is a fairly special case, but overall the story with this is "nobody looked at it" - I tried to start a discussion on what it would take to make it work together with granite, but you just went with ignoring `adw_init()` (which is also unsupported and works by accident btw, and you still miss stuff like translations) :p

Edit: I've asked on slack for a clarification, would be nice to get this going. :)

**Reply by alice-mkh** [reply]
Ok, so after a bit of research: we can detect Granite presence on libadwaita side with a bit of dlsym magic and disable style manager stuff. However:

- It needs ephy to link to Granite
- It needs ephy to actually use Granite symbols

We could probably have it as a compile time option. For example, off by default, then elementary flips it to true and maybe we do as well for flathub and nightly builds (not too much of a stretch given we already bundle elementary icons in both, and granite is much smaller than that).

Now, we don't actually use anything from Granite. So on one hand this becomes trickier, but on the other hand it means we could do something (can be a bogus call to something like `granite_accel_to_string()`) conditionally depending on `XDG_CURRENT_DESKTOP` to have it run correctly in both like the gtk3 version.

We'd also need to have an elementary version of app-provided styles but we really should do it anyway - we had a lot of adjustments to make it look nice with the old elementary stylesheet, but that stopped working in 6.0 since it was renamed from `elementary` to `io.elementary.stylesheet.<accent>`. But if we do what I'm describing, we can just assume that if elementary integration is enabled and we are on pantheon - it won't be adwaita and we can load those styles instead.

Mainly a question to @mcatanzaro: do we want to go that route?

**Reply by mcatanzaro** [reply]
I'm OK with that.

**Reply by alice-mkh** [reply]
OK, libadwaita side is up: https://gitlab.gnome.org/GNOME/libadwaita/-/merge_requests/768

Ephy side will take more time, because there's apparently a Vala bug with `gtk_css_provider_load_from_data()` and Granite doesn't build atm.

**Reply by alice-mkh** [reply]
Aand there's a hole in my plan: using a Granite symbol period will cause it to be loaded if it's if()-ed away.

So that doesn't work, we need some better check. Ideas?

**Reply by mcatanzaro** [reply]
Why not just check if `$XDG_CURRENT_DESKTOP ` contains `Pantheon`...?

**Reply by alice-mkh** [reply]
On libadwaita side? Because that would also contain Pantheon for other apps that aren't Epiphany or elementary apps using libadwaita.

**Reply by Marukesu** [reply]
@Exalm would be having a programmer-provided style manager possible? if so, we could have a Granite.StyleManager that epiphany can use.

otherwise, a explict `adw_disable_styles()` function would be the simple route, i think.

**Reply by alice-mkh** [reply]
> @Exalm would be having a programmer-provided style manager possible? if so, we could have a Granite.StyleManager that epiphany can use.

That's already a non-issue. The issue is reliably detecting granite, see #browser on slack.

> otherwise, a explict adw_disable_styles() function would be the simple route, i think.

Not happening, it's a "break me" option when used outside granite.

**Reply by alice-mkh** [reply]
Ephy side is up as well: https://gitlab.gnome.org/GNOME/epiphany/-/merge_requests/1292

**Comment by alice-mkh** [comment]
Ok, everything is merged - the latest nightly is using elementary styles.

![Screenshot from 2023-03-03 23-26-14](https://user-images.githubusercontent.com/1510457/222809057-b4d8d92e-3035-4e28-8b6e-2de6a85bfe87.png)

Now there are a bunch of things to fix on your side:

- libadwaita disables its own icons in Granite compat mode, so you will need to provide the missing icons yourself

![Screenshot from 2023-03-03 23-27-09](https://user-images.githubusercontent.com/1510457/222809223-ab88ea6e-fc25-4d9a-8f73-da7bbb405f95.png)

![Screenshot from 2023-03-03 23-27-33](https://user-images.githubusercontent.com/1510457/222809278-adbd5d39-ff23-482a-8a22-a0d779b8f0ba.png)

![Screenshot from 2023-03-03 23-27-51](https://user-images.githubusercontent.com/1510457/222809326-437811da-7a2c-4efc-91f6-167b2b867bdf.png)

![Screenshot from 2023-03-03 23-36-10](https://user-images.githubusercontent.com/1510457/222810790-1333c564-7a51-422f-9c24-a40baaa19402.png)


- there are _tons_ of missing styles

![Screenshot from 2023-03-03 23-28-12](https://user-images.githubusercontent.com/1510457/222809389-e94b7bf3-136b-4de3-abe0-4b47b7a8a9ae.png)

`.boxed-list`, libadwaita rows, `AdwPreferencesPage`, `AdwPreferencesGroup`

![Screenshot from 2023-03-03 23-28-34](https://user-images.githubusercontent.com/1510457/222809574-62940da0-bce3-45c0-98f1-1ab5435f0bc1.png)

`GtkSearchBar`

![Screenshot from 2023-03-03 23-28-52](https://user-images.githubusercontent.com/1510457/222809505-5973c566-bbaf-49bf-8aa2-b70701c43640.png)

Listboxes inside `popover.menu`

![Screenshot from 2023-03-03 23-29-07](https://user-images.githubusercontent.com/1510457/222809554-deac99e6-639c-4b3e-a9bc-c519bcf76802.png)

A part of this one is an X11-specific gtk bug, but still missing `AdwStatusPage` styles.

Also note the missing spacing between the bookmarks and settings buttons - the rest of the buttons are in boxes with hardcoded spacing, but not these 2, so looks like header bars are missing spacing in general

![Screenshot from 2023-03-03 23-32-09](https://user-images.githubusercontent.com/1510457/222810126-c125e0fe-61e2-4727-ab6a-b2f70048e7dc.png)

Looks like there is some weird spacing between header bar and tab bar? At least in fullscreen

![Screenshot from 2023-03-03 23-38-18](https://user-images.githubusercontent.com/1510457/222811170-2599eb0e-efcb-4c71-947c-36039248344a.png)

Header bar shadows end up behind content and can't really work here. We'll have a widget to make them work on libadwaita side in future, but obviously not for 44.

- the clipped labels in tabs are because of missing gtk-hint-font-metrics=1 - nothing I can do there, the clip bounds are correct on the libadwaita side

There are also missing styles upstream, e.g. checked state for reader mode icon or incognito styles. MRs welcome, though at this point they will be for 45.

**Reply by alice-mkh** [reply]
(Ignore the broken shadows on buttons - that's specific to the VM and it looks fine outside it here:

![image](https://user-images.githubusercontent.com/1510457/222812218-ac003288-7dcd-4033-9036-c3320bae4922.png)

**Reply by danirabbit** [reply]
Fantastic work thank you! I’ll get those all fixed up for sure
